### PR TITLE
More fixes

### PR DESCRIPTION
**Pull request by AlexDBlack**
Fixes:
- https://github.com/deeplearning4j/deeplearning4j/issues/3594
- https://github.com/deeplearning4j/deeplearning4j/issues/3493 + adds multi-IEvaluation support

Implements:
- https://github.com/deeplearning4j/deeplearning4j/issues/3616

### CODE REVIEW

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/graph/ComputationGraph.java [hunk 1]
```diff
@@ -2953,7 +2953,7 @@ public ROCMultiClass evaluateROCMultiClass(MultiDataSetIterator iterator, int ro
      * Perform evaluation on the given data (MultiDataSetIterator) with the given {@link IEvaluation} instance
      *
      * @param iterator   Test data to evaluate on
-     * @param evaluation IEvaluation insntance
+     * @param evaluations IEvaluation insntance
```
Comment (huitseeker) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/graph/ComputationGraph.java]: Typo

File: deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/graph/SparkComputationGraph.java [hunk 2]
```diff
@@ -632,10 +696,43 @@ public Evaluation evaluate(JavaRDD<DataSet> data, List<String> labelsList, int e
...
      */
+    @SuppressWarnings("unchecked")
     public <T extends IEvaluation> T doEvaluation(JavaRDD<DataSet> data, T emptyEvaluation, int evalBatchSize) {
+        IEvaluation[] arr = new IEvaluation[]{emptyEvaluation};
+        return (T)doEvaluation(data, evalBatchSize, arr)[0];
+    }
+
+    /**
+     * Perform distributed evaluation on a <i>single output</i> ComputationGraph form DataSet objects using Spark.
+     * Can be used to perform multiple evaluations on this single output (for example, {@link Evaluation} and
```
Comment (huitseeker) [inline deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/graph/SparkComputationGraph.java]: is your 1st link what you intended ? I'd have expected something more specific.

File: deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/multilayer/SparkDl4jMultiLayer.java [hunk 3]
```diff
@@ -574,9 +575,23 @@ public Evaluation evaluate(JavaRDD<DataSet> data, List<String> labelsList, int e
...
+     * @param <T>              Type of evaluation instance to return
+     * @return IEvaluation instances
+     */
+    public <T extends IEvaluation> T[] doEvaluation(JavaRDD<DataSet> data, int evalBatchSize, T... emptyEvaluations) {
         IEvaluateFlatMapFunction<T> evalFn = new IEvaluateFlatMapFunction<>(false, sc.broadcast(conf.toJson()),
-                        sc.broadcast(network.params()), evalBatchSize, emptyEvaluation);
-        JavaRDD<T> evaluations = data.mapPartitions(evalFn);
+                sc.broadcast(network.params()), evalBatchSize, emptyEvaluations);
+        JavaRDD<T[]> evaluations = data.mapPartitions(evalFn);
         return evaluations.reduce(new IEvaluationReduceFunction<T>());
```
Comment (huitseeker) [inline deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/multilayer/SparkDl4jMultiLayer.java]: `reduce(mapPartitions)` is very often `RDD.aggregate()`, If we can reformulate this in the following terms, we get perf goodies like `treeAggregate` nearly for free: 
```
def aggregate[U](zeroValue: U)(seqOp: (U, T) ⇒ U, combOp: (U, U) ⇒ U): U 
```
Just a thought, dunno if it's worth it at this stage.

File: deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/graph/SparkComputationGraph.java (hunk 2 above)
Comment (AlexDBlack) [inline deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/graph/SparkComputationGraph.java]: Link shows up correctly once rendered in javadoc (due to import of Evaluation class)

File: deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/multilayer/SparkDl4jMultiLayer.java (hunk 3 above)
Comment (AlexDBlack) [inline deeplearning4j-scaleout/spark/dl4j-spark/src/main/java/org/deeplearning4j/spark/impl/multilayer/SparkDl4jMultiLayer.java]: Thanks, I changed this to .treeAggregate()

**Pull request merged**